[WIP] More flexible and convenient logging

c42f opened the pull request:
Here's a WIP for making julia logging systematic, flexible and fast, as described in part in [the logging julep](https://github.com/JuliaLang/Juleps/pull/30), and currently being prototyped in [MicroLogging.jl](https://github.com/c42f/MicroLogging.jl). It turns out that there's quite a few disparate mechanisms in use already for logging in Base, so this WIP is about looking at each use case in turn. Attacking `depwarn` has been particularly entertaining, and I think the result so far is faster and significantly more accurate. As a side effect of this work, I think we should be easily able to generate a depwarn report after any given julia run with a precise list of locations to fix.

As a whole this is probably not that close to merging yet, but a lot of the key pieces are in place.  There's a design tracker over at https://github.com/c42f/MicroLogging.jl/issues/5 with more detail on unanswered issues.

I've included only a bare minimum of log formatting machinery here with the idea that this is better done in a standard library package.  With that in mind, I'd suggest checking out the  `MicroLogging` master branch for logs which are actually formatted in a readable way for interactive work.

c42f committed:
Copy of MicroLogging source (656f937) exports + sysimg

Some minor edits here to make things suitable for Base.

c42f committed:
Convert Pkg to use new Logging

c42f committed:
Convert libgit2 to use new logging

c42f committed:
Migrate depwarn to new logging infrastructure

* Forward depwarn() calls to new backend
* Use caller frame pointer as message id.  This makes initial
  first-pass deduplication of depwarns quite a lot faster.
* Change --depwarn=error to --depwarn=throw for clarity
* Introduce defines for log levels into julia.h

ggggggggg commented on base/logging.jl:
```diff
@@ -0,0 +1,509 @@
+# itself.
+struct LogState
+    min_enabled_level::LogLevel
+    logger::AbstractLogger
```
Isn't this not concretely typed?

c42f replied:
`LogState` itself is a leaf type; `AbstractLogger` is not. This is arranged so that some minimal part of the state (only `min_enabled_level` at the moment, but possibly some additional things) is accessible without generic dispatch.

ggggggggg replied:
I think I get it, you concretely typed as much as possible, but logger can't be concretely typed if you want to be able to change it easily?

c42f replied:
Exactly.  The design problem is to decide which aspects of log filtering are sufficiently useful and fast that every user defined logger must support them - those can be cached in the concretely typed log state, and inlined at the call site.

ggggggggg commented on base/logging.jl:
```diff
@@ -0,0 +1,509 @@
+const _min_enabled_level = Ref(Debug)
+
+
+# A concretely typed cache of data extracted from the logger, plus the logger
```
Most of these comments look like they should be docstrings.

c42f replied:
The convention I'm using is that implementation details have comments but no docstrings. How is this generally handled in `Base`? I don't want to give people the idea that `Logging.LogState` (for example) is part of the public API.

ggggggggg replied:
That seems sensible. A quick browse shows me that there is a non-exported function in Enums.jl with a comment rather than a doc-string, so it seems similar to what you're doing.

c42f committed:
Forward all frontend depwarns to the logging system

Forward all frontend depwarn messages through to the julia logging
system for consistency of formatting and dispatch.  Deciding whether a
depwarn should appear as an thrown error is also now handled outside
flisp to keep this decision in one place where possible.

Detail:

* Ensure that deprecation-message forwards to the logging system.  Also
  include file and line number metadata
* Remove jl_parse_depwarn(), flisp *depwarn* and related scheme code as
  this can be handled on the julia side.
* Rename syntax_deprecation_warnings(false) to
  without_syntax_deprecations, as this was only ever used to turn syntax
  deprecations off.
* Ensure that all lowering depwarns with existing line number
  information get this passed through as accessible metadata rather than
  as a string.
* Use distinct functions for depwarns coming from the parser vs
  lowering, as these get line number information in a different way.

TODO:
* Figure out a decent `id` for the syntax and lowering depwarn messages
* Make without_syntax_deprecations filter only depwarn messages, rather
  than everything!

c42f committed:
Migrate loading to new logging infrastructure

c42f commented:
Hmm, mysterious - it looks like #17997 broke something I was doing on this branch. I'm now getting odd errors regarding `is_root_module` during bootstrap even after rebasing locally and fixing some conflicts.

c42f committed:
More backend interface renaming

Use handle_message rather than dispatch_message, as not all message
handling is pure dispatch (the message may be stored rather than sent
elsewhere).

Reuse the name dispatch_message internally.

c42f committed:
Changes ported from MicroLogging

* Additional exports
* Fallback error message formatting
* Minor global logger init min level bug

c42f committed:
Quote depwarns code expressions with backticks

helgee commented on base/libgit2/merge.jl:
```diff
@@ -219,13 +219,13 @@ function merge!(repo::GitRepo, anns::Vector{GitAnnotated}, fastforward::Bool;
         if isset(ma, Cint(Consts.MERGE_ANALYSIS_FASTFORWARD))
             if length(anns) > 1
-                warn("Unable to perform Fast-Forward merge with mith multiple merge heads.")
+                @warn "Unable to perform Fast-Forward merge with mith multiple merge heads." repo
```
Not your fault but "with mith"? 😂

pkofod replied:
s/with mith/mit

;) :de:

c42f replied:
Fixed!

c42f committed:
libgit2 logging - minor fixes and updates

c42f committed:
Pkg logging - exception and markdown fixes

c42f committed:
Remove sprintf dependency - use hex() and string() instead.

c42f committed:
Upstream changes: docs + keyword splatting

c42f committed:
Move core Logging components into Base

For now this seems like the right thing as error and informational
reporting is a core requirement for a compiler - parts of the frontend
and runtime now call the logging system when it's available.

c42f committed:
Tweak loading for core Logging in Base

c42f committed:
Tweak depwarns for core Logging in Base

c42f committed:
Logging tools for use in the runtime C code

These can be used to transform a lot of printf()'s into log statements.

c42f committed:
Add catch_exceptions(::AbstractLogger)

This allows users to control whether a custom logger type will catch
exceptions during message creation and log them, or simply allow them to
pass through.

c42f committed:
Deprecate --depwarn=error - use a custom logger instead

c42f committed:
Add module argument to jl_log() and jl_logf()

c42f committed:
Implement general julia-logmsg function for flisp logging

Provides a general tool to log information from the flisp frontend code.
Make the syntax deprecations use this instead of using the purpose built
syntax_depwarn function on the julia side.

Also reinstate flisp-level awareness of the depwarn level.  It's just
simpler.

c42f committed:
Change log level limits to fit well within a flisp fixnum

This is purely for implementation convenience.  It restricts the levels
artifically, but it's hard to imagine what anyone would do with more than
a million custom levels.

c42f committed:
Remove all uses of --depwarn=error

This can be handled in a much nicer way by installing a logger in-process.

TODO: Define such a logger and test it!

c42f committed:
Cleanup: Extend Base methods with explicit Base prefix

StefanKarpinski commented:
Should we get some review on this?

c42f commented:
Some high level feedback would be great.  Particularly around how the system supports the various logging needs of base.

Just don't be too disturbed if there's a few details which are still broken here - it's definitely still a WIP, and I plan to pull it apart into several pieces for detailed review.

c42f committed:
Rearrange logging source into a more logical order

Port from MicroLogging refactoring

Also rearrange exports to avoid exporting AbstractLogger-related methods.

c42f committed:
Port more MicroLogging changes - remove log config from core

ViralBShah commented:
cc @tanmaykm

c42f commented:
Ok, I've pulled out what I think is a mostly coherent subset of these changes into #24490 .  This WIP is still useful for the moment as context for the larger ongoing work.

c42f commented:
This is superseded now by #24490 and various follow on work.
